File: webstreamjar/app/src/main/kotlin/ru/simlabs/stream/unreal/AndroidMediaTextureUpdater.kt
package ru.simlabs.stream.unreal

import android.view.Surface
import ru.simlabs.stream.StreamDecoder

class AndroidMediaTextureUpdater(private val width: Int, private val height: Int) : ()->StreamDecoder {
    private val bitmapRenderer: BitmapRenderer = BitmapRenderer(false)
    private var textureID: Int = -1

    init {
        bitmapRenderer.setSize(width, height)
    }

    fun setTextureID(textureID: Int) {
        this.textureID = textureID
    }

    override fun invoke(): StreamDecoder {
        return StreamDecoder(false, bitmapRenderer.surface, width, height)
    }

    fun updateTexture() {
        if (textureID != -1) {
            bitmapRenderer.updateFrameData(textureID)
        }
    }
}

File: SimLabsVR/Source/SimLabsVR/Private/Android/AndroidMediaTextureUpdater.cpp
#include "Android/AndroidMediaTextureUpdater.h"
#include "Engine.h"
#include "RenderingThread.h"

namespace SimlabsStream
{

#if PLATFORM_ANDROID
IMediaTextureUpdater *CreateMediaTextureUpdater(UMediaTexture *MediaTexture, int width, int height)
{
	auto* TextureUpdater = new FAndroidMediaTextureUpdater(width, height);
	TextureUpdater->InitTexture(MediaTexture);
	return TextureUpdater;
}
#endif

FAndroidMediaTextureUpdater::FAndroidMediaTextureUpdater(int width, int height)
	: FJavaClassObject(FName("ru/simlabs/stream/unreal/AndroidMediaTextureUpdater"), "(II)V", width, height)
	, UpdateTextureMethod(GetClassMethod("updateTexture", "()V"))
	, SetTextureIDMethod(GetClassMethod("setTextureID", "(I)V"))
	, textureWidth(width)
	, textureHeight(height)
{
}

void FAndroidMediaTextureUpdater::InitTexture(UMediaTexture *MediaTexture)
{
	ENQUEUE_UNIQUE_RENDER_COMMAND_TWOPARAMETER(SimlabsStreamAndroidDecoderInitTexture,
		UMediaTexture*, Texture, MediaTexture,
		FAndroidMediaTextureUpdater*, This, this,
		{
			Texture->InitializeTextureSink(
				{ This->textureWidth, This->textureHeight },
				{ This->textureWidth, This->textureHeight },
				EMediaTextureSinkFormat::CharBGRA,
				EMediaTextureSinkMode::Unbuffered
			);
			int textureResource = *reinterpret_cast<int*>(Texture->GetTextureSinkTexture()->GetNativeResource());
			This->SetTextureID(textureResource);
		}
	);
}

void FAndroidMediaTextureUpdater::SetTextureID(int textureID) {
	CallMethod<void>(SetTextureIDMethod, textureID);
}

void FAndroidMediaTextureUpdater::UpdateTexture() {
	ENQUEUE_UNIQUE_RENDER_COMMAND_ONEPARAMETER(SimlabsStreamAndroidDecoderUpdateTexture,
		FAndroidMediaTextureUpdater*, This, this,
		{
			This->CallMethod<void>(This->UpdateTextureMethod);
		}
	);
}

}